cps/services/oauth.py

```diff
@@ -0,0 +1,327 @@
+# -*- coding: utf-8 -*-
+
+#  This file is part of the Calibre-Web (https://github.com/janeczku/calibre-web)
+#    Copyright (C) 2018-2019 OzzieIsaacs, jim3ma, pwr
+#
+#  Licensed under GLPv3. See the project's LICENSE file for details.
+
+from __future__ import division, print_function, unicode_literals
+
+from flask import session as flask_session
+from flask_dance import __version__ as flask_dance_version
+from flask_dance.consumer import oauth_authorized, oauth_error
+from flask_dance.consumer.storage.sqla import SQLAlchemyStorage
+from flask_dance.contrib.github import make_github_blueprint
+from flask_dance.contrib.google import make_google_blueprint
+from sqlalchemy.orm.exc import NoResultFound, MultipleResultsFound
+
+from .. import logger
+
+
+log = logger.create()
+providers = {}
+
+_SUPPORTED = {
+    'github': ('GitHub', make_github_blueprint,
+               'https://github.com/settings/developers', '/user'),
+    'google': ('Google', make_google_blueprint,
+               'https://console.developers.google.com/apis/credentials', '/oauth2/v2/userinfo')
+}
+
+
+class _Provider(object):
+    '''
+    Wrapper over an external OAuth provider.
+    Works with the current flask session to store/cache some of the state.
+    '''
+    def __init__(self, id_, redirect_to):
+        self.id = id_
+
+        name, make_oauth_blueprint, obtain_link, account_info_path = _SUPPORTED[id_]
+        self.blueprint = make_oauth_blueprint(storage=_STORAGE)
+        assert id_ == self.blueprint.name, "provider '%s' should have been named '%s'" % (id_, self.blueprint.name)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846306)